x-pack/plugins/security_solution/public/resolver/models/schema.ts

```diff
@@ -0,0 +1,176 @@
+ * This allows us to write object literals that omit a key if the value can accept `undefined`.
+ */
+type OptionalKeyWhenValueAcceptsUndefined<T extends { [key: string]: unknown }> = {
+  [K in Exclude<keyof T, KeysWithOptionalValues<T>>]: T[K];
+} &
+  {
+    [K in KeysWithOptionalValues<T>]?: Exclude<T[K], undefined>;
+  };
+
+/**
+ * Validate that `value` is an object with string keys. The value at each key is tested against its own validator.
+ *
+ * Use this to create a predicate for a type like `{ a: string[] }`. For example:
+ * ```ts
+ * import * as schema from './schema';
+ * const myValidator: (value: unknown) => value is { a: string[] } = schema.object({
+ *   a: schema.array(schema.string()),
+ * });
+ * ```
+ */
+export function object<
+  ValidatorDictionary extends {
+    [key: string]: Validator<unknown>;
+  }
+>(validatorDictionary: ValidatorDictionary) {
+  return function (
+    value: unknown
+  ): value is /** If a key can point to `undefined`, then instead make the key optional and exclude `undefined` from the value type. */ OptionalKeyWhenValueAcceptsUndefined<
+    {
+      [K in keyof ValidatorDictionary]: TypeOf<ValidatorDictionary[K]>;
+    }
+  > {
+    // This only validates non-null objects
+    if (typeof value !== 'object' || value === null) {
+      return false;
+    }
+
+    // Rebind value as the result type so that we can interrogate it
+    const trusted = value as { [K in keyof ValidatorDictionary]: TypeOf<ValidatorDictionary[K]> };
+
+    // Get each validator in the validator dictionary and use it to validate the corresponding value
+    for (const key of Object.keys(validatorDictionary)) {
+      const validator = validatorDictionary[key];
+      if (!validator(trusted[key])) {
+        return false;
+      }
+    }
+    return true;
+  };
+}
+
+/**
+ * Validate that `value` is strictly equal to `acceptedValue`.
+ * Use this for a literal type, for example:
+ * ```
+ * import * as schema from './schema';
+ * const isAscOrDesc: (value: unknown) => value is 'asc' | 'desc' = schema.oneOf([
+ *   schema.literal('asc' as const),
+ *   schema.literal('desc' as const),
+ * ]);
+ * ```
+ */
+export function literal<T>(acceptedValue: T) {
+  return function (value: unknown): value is T {
+    return acceptedValue === value;
+  };
+}
+
+/**
+ * Validate that `value` is a string.
+ * NB: this is used as `string` externally via named export.
+ * Usage:
+ * ```
+ * import * as schema from './schema';
+ * const isString: (value: unknown) => value is string = schema.string();
+ * ```
+ */
+function anyString(): (value: unknown) => value is string {
+  return function (value: unknown): value is string {
+    return typeof value === 'string';
+  };
+}
+
+/**
+ * Validate that `value` is a number.
+ * NB: this just checks if `typeof value === 'number'`. It will return `true` for `NaN`.
```

Comment:
❔ 
> It will return `true` for `NaN`.
Is that OK? Are there repercussions to that? Should the function be named differently (i.e. Would you be surprised to get something from a function called `anyNumber` that is, by definition `Not a Number`?) `anyNumberOrNan`, `anyTypeOfNumber`

Reply:
Or you could use like `Number.isFinite` to catch other weird stuff that lives in `typeof === "number"` like `Infinity` and call it `anyFiniteNumber`

Reply:
It's worth discussing.